crates/sui-framework/packages/dwallet-system/sources/sui_state_proof.move

```diff
@@ -0,0 +1,182 @@
+#[allow(unused_field)]
+module dwallet_system::sui_state_proof {
+
+    use dwallet::object::{Self, ID, UID};
+    use dwallet_system::dwallet::{Self, DWalletCap, MessageApproval};
+    use dwallet::tx_context::{TxContext};
+    use dwallet::transfer;
+    use dwallet::bcs;
+    use dwallet::event;
+    use std::vector;
+
+
+    const EWrongEpochSubmitted: u64 = 0;
+    const EWrongDWalletCapId: u64 = 1;
+    const EStateProofNoMessagesToApprove: u64 = 2;
+
+
+    struct StateProofRegistry has key, store {
+            id: UID,
+            highest_epoch: u64,
+    }
+
+    struct StateProofConfig has key, store{
+        id: UID,
+        registry_id: ID,
+        package_id: vector<u8>,
+        init_cap_event_type_layout: vector<u8>,
+        approve_event_type_layout: vector<u8>,
+    }
+
+    struct EpochCommittee has key, store {
+        id: UID,
+        committee: vector<u8>,
+    }
+
+
+    struct EpochCommitteeSubmitted has copy, drop {
+        epoch: u64,
+        epoch_committee_id: ID,
+    }
+
+    struct CapWrapper has key, store {
+        id: UID,
+        cap_id_sui: ID,
+        cap: DWalletCap,
+    }
+
+
+    native fun sui_state_proof_verify_committee(prev_committee: vector<u8>, checkpoint_summary: vector<u8>): (vector<u8>, u64);
+
+    native fun sui_state_proof_verify_link_cap(committee: vector<u8>, checkpoint_summary: vector<u8>, checkpoint_contents: vector<u8>, transaction: vector<u8>,  event_type_layout: vector<u8>,  package_id: vector<u8>): (vector<u8>, vector<u8>);
+
+    native fun sui_state_proof_verify_transaction(committee: vector<u8>, checkpoint_summary: vector<u8>, checkpoint_contents: vector<u8>, transaction: vector<u8>,  event_type_layout: vector<u8>,  package_id: vector<u8>): (vector<u8>, vector<u8>);
+
+
+
+    public fun init_module(
+        init_committee: vector<u8>, 
+        package_id: vector<u8>, 
+        init_cap_event_type_layout: vector<u8>, 
+        approve_event_type_layout: vector<u8>, 
+        epoch_id_committee: u64, 
+        ctx: &mut TxContext) {
+        let registry = StateProofRegistry {
+            id: object::new(ctx),
+            highest_epoch: epoch_id_committee,
+        };
+
+        let config = StateProofConfig {
+            id: object::new(ctx),
+            registry_id: object::id(&registry),
+            package_id: package_id,
+            init_cap_event_type_layout: init_cap_event_type_layout,
+            approve_event_type_layout: approve_event_type_layout,
+        };
+
+        let first_committee = EpochCommittee {
+            id: object::new(ctx),
+            committee: init_committee,
+        };
```

Comment:
I'm wondering if there could be a sanity check here that the committee is of the right structure. In any case, this function is an authorized one and users should check it was initialized correctly, so this isn't a big deal security wise (e.g. if it was attempted to be initialized maliciously one could just put a wrong committee) but if we could at least make sure that this deserializes into the correct structure it could be good